torch/distributed/_shard/replicated_tensor.py

```diff
@@ -5,6 +5,14 @@
 from torch.distributed import distributed_c10d
 from torch.overrides import get_default_nowrap_functions
 
+_REPLICATED_WITH_NON_TENSOR_ALLOWLIST = [
```

Comment:
I think this list is for the case where all args are either ReplicatedTensor or scalars, instead of having two flags `all_replicated` and `replicated_with_non_tensor`, could we have a single flag to check if there's no `torch.Tensor` in the args (ShardedTensor or PartialTensor handles in their own __torch_function__), then we convert to ReplicatedTensor, otherwise leave it as tensor?

Reply:
Hmm but couldn't there be a case where there is an op with ReplicatedTensor and PartialTensor and it gets dispatched to ReplicatedTensor's `torch_function` instead? We re-dispatch to ShardedTensor in the case of ShardedTensor, but I don't see anything similar for ReplicatedTensor.

Reply:
So another issue is Replicated (op) scalar is not always guaranteed to be ReplicatedTensor. Imagine torch.add(ReplicatedTensor, x) where `x` is different on each rank.

So it seems like we do need this additional flag in which case we have look through a whitelist like this.